File: superstitioApi/java/superstitioapi/powers/barIndepend/BarRenderUpdateMessage.java
package superstitioapi.powers.barIndepend;

import com.badlogic.gdx.graphics.Color;
import com.megacrit.cardcrawl.helpers.PowerTip;

import java.util.function.Consumer;

public class BarRenderUpdateMessage {
    Color chunkColor;
    String uuidOfPower;
    String uuidOfBar;
    int newAmount;
    Consumer<RenderOnThing> detail;
    int maxAmount;
    String rawTextOnBar;
    ToolTip toolTip;

    public BarRenderUpdateMessage(String uuidOfPower, String uuidOfBar) {
        this.uuidOfPower = uuidOfPower;
        this.uuidOfBar = uuidOfBar;
    }

    public PowerTip toPowerTip(ToolTip toolTip) {
        return new PowerTip(toolTip.name, toolTip.description);
    }

    public BarRenderUpdateMessage setRawTextOnBar(String rawTextOnBar) {
        this.rawTextOnBar = rawTextOnBar;
        return this;
    }

    public BarRenderUpdateMessage setMaxAmount(int maxAmount) {
        this.maxAmount = maxAmount;
        return this;
    }

    public BarRenderUpdateMessage setNewAmount(int newAmount) {
        this.newAmount = newAmount;
        return this;
    }

    public BarRenderUpdateMessage setDetail(Consumer<RenderOnThing> detail) {
        this.detail = detail;
        return this;
    }

    public BarRenderUpdateMessage setTip(ToolTip toolTip) {
        this.toolTip = new ToolTip(toolTip);
        return this;
    }

    public BarRenderUpdateMessage setChunkColor(Color newColor) {
        this.chunkColor = newColor.cpy();
        return this;
    }

    public static class ToolTip {
        String name;
        String description;

        public ToolTip(String name, String description) {
            this.name = name;
            this.description = description;
        }

        public ToolTip(ToolTip toolTip) {
            this.name = toolTip.name;
            this.description = toolTip.description;
        }

    }
}

File: superstitioApi/java/superstitioapi/powers/barIndepend/BarRenderUpdateMessage.kt
package superstitioapi.powers.barIndepend

import com.badlogic.gdx.graphics.Color
import com.megacrit.cardcrawl.helpers.PowerTip
import java.util.function.Consumer

class BarRenderUpdateMessage(var uuidOfPower: String, var uuidOfBar: String)
{
    var chunkColor: Color? = null
    var newAmount: Int = 0
    var detail: Consumer<RenderOnThing>? = null
    var maxAmount: Int = 0
    var rawTextOnBar: String? = null
    var toolTip: ToolTip? = null

    fun toPowerTip(toolTip: ToolTip): PowerTip
    {
        return PowerTip(toolTip.name, toolTip.description)
    }

    fun setRawTextOnBar(rawTextOnBar: String?): BarRenderUpdateMessage
    {
        this.rawTextOnBar = rawTextOnBar
        return this
    }

    fun setMaxAmount(maxAmount: Int): BarRenderUpdateMessage
    {
        this.maxAmount = maxAmount
        return this
    }

    fun setNewAmount(newAmount: Int): BarRenderUpdateMessage
    {
        this.newAmount = newAmount
        return this
    }

    fun setDetail(detail: Consumer<RenderOnThing>?): BarRenderUpdateMessage
    {
        this.detail = detail
        return this
    }

    fun setTip(toolTip: ToolTip): BarRenderUpdateMessage
    {
        this.toolTip = ToolTip(toolTip)
        return this
    }

    fun setChunkColor(newColor: Color): BarRenderUpdateMessage
    {
        this.chunkColor = newColor.cpy()
        return this
    }

    class ToolTip
    {
        var name: String
        var description: String

        constructor(name: String, description: String)
        {
            this.name = name
            this.description = description
        }

        constructor(toolTip: ToolTip)
        {
            this.name = toolTip.name
            this.description = toolTip.description
        }
    }
}
